src/utils.js

```diff
@@ -234,3 +257,189 @@ export function trimContentDam(path) {
+}
+
+/**
+ * Determines whether or not a given path is either a temp file, or in a temp directory.
+ * @param {string} path A file system-like path.
+ * @returns {boolean} True if the path is a temp file, false otherwise.
+ */
+export function isTempPath(path) {
+    const tempPath = normalizePath(path);
+
+    if (tempPath === '/') {
+        return false;
+    }
+
+    let isTemp = TEMP_PATTERNS.some(pattern => pattern.test(tempPath));
+
+    if (!isTemp) {
+        const pathName = Path.basename(tempPath);
+        isTemp = TEMP_NAME_PATTERNS.some(pattern => pattern.test(pathName));
+    }
+
+    return isTemp;
+}
+
+/**
+ * Concurrently loops through all items in a directory, doing a stat
+ * on each to determine if it's a directory or file. The method will
+ * skip temp files and directories.
+ * @param {string} directoryPath Full path to the directory to iterate.
+ * @returns {object} Simple object containing the following elements:
+ *  {Array} directories: Full path to all directories beneath the given
+ *   directory. The list is not guaranteed to be sorted, but it's guaranteed
+ *   that parent directories will always come before their sub directories.
+ *  {Array} files: Full path to all files beneath the given directory. The list
+ *   is not guaranteed to be sorted, but it's guaranteed that files higher in
+ *   the directory hierarchy will always come before those lower in the hierarchy.
+ *  {number} totalSize: Total size, in bytes, of all files in the directory.
+ */
+async function processDirectory(directoryPath) {
+    let contents = false;
+    
+    const directories = [];
+    const files = [];
+    const errors = [];
+    let totalSize = 0;
+    let includeDirectory = false;
+
+    try {
+        contents = await fs.readdir(directoryPath);
+    } catch (e) {
+        errors.push(e);
+    }
+
+    if (contents) {
+        await concurrentLoop(contents, async (childPath) => {
+            const fullChildPath = Path.join(directoryPath, childPath);
+            if (!isTempPath(fullChildPath)) {
+                let childStat;
+                try {
+                    childStat = await fs.stat(fullChildPath);
+                } catch (e) {
+                    errors.push(e);
+                    return;
+                }
+
+                if (childStat.isDirectory()) {
+                    directories.push({ path: fullChildPath });
+                    includeDirectory = true;
+                } else if (childStat.isFile()) {
+                    files.push({ path: fullChildPath, size: childStat.size });
+                    totalSize += childStat.size;
+                    includeDirectory = true;
+                }
+            }
+        });
+    }
+
+    return {
+        directories,
+        files,
+        errors,
+        totalSize,
+        includeDirectory
+    };
+}
+
+function removeFromArray(valueToRemove, array) {
+    for (let i = 0; i < array.length; i++) {
+        if (array[i].path === valueToRemove) {
```

Comment:
Looking at this, I think using a Map would be better.
See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map: it's a key => value pair. The Map would avoid having a lot of objects moving around (no splice needed).
Wdyt?

Reply:
(Maps also have a size, so the algorithm would be similar to this)

Reply:
Thanks for the callout. I did some refactoring to simplify the function in general, and was able to get rid of `removeFromArray`.